File: leetcode/0169_majority_element.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <cassert>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
    int majorityElement(vector<int>& nums) {
        unordered_map<int, int> counts;
        int majority = nums.size() / 2 + 1;

        for (const auto& num : nums) {
            counts[num]++;
            if (counts[num] >= majority) {
                return num;
            }
        }
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {3, 2, 3};

    assert(3 == solution.majorityElement(nums));
}

File: leetcode/0169_majority_element.py
# -*- coding: utf-8 -*-

from collections import Counter


class Solution(object):
    def majorityElement(self, nums):
        return Counter(nums).most_common(1)[0][0]


if __name__ == '__main__':
    solution = Solution()

    assert 1 == solution.majorityElement([1])
    assert 1 == solution.majorityElement([1, 1, 2])
